File: ios-client/Workrec/ApiClient/ApiClient.swift
//
//  ApiClient.swift
//  Workrec
//
//  Created by ishida on 2023/03/29.
//

import Apollo
import Combine
import FirebaseAuth
import Foundation
import WorkrecGraphql

struct ApiError: Error {
  let message: String
}

enum TaskState: String {
  case notStarted = "not_started"
  case inProgress = "in_progress"
  case paused = "paused"
  case completed
}

struct TaskListItem: Identifiable {
  let id: String
  let title: String
  let state: TaskState
  let totalWorkingTime: Int
  let lastStartTime: Date

  var totalWorkingTimeInCurrent: Int {
    if state == .inProgress {
      return totalWorkingTime + Int((Date.now.timeIntervalSince(lastStartTime)))
    }
    return totalWorkingTime
  }
}

class ApiClient: ObservableObject {
  private var apolloClient: ApolloClient! = nil

  @Published var authToken: String = ""

  init() {
    let cache = InMemoryNormalizedCache()
    let store = ApolloStore(cache: cache)

    let client = URLSessionClient()
    let provider = NetworkInterceptorProvider(client: client, store: store, provideAuthToken: { self.authToken })
    let url = URL(string: "http://localhost:8080/graphql")!

    let requestChainTransport = RequestChainNetworkTransport(
      interceptorProvider: provider,
      endpointURL: url
    )

    self.apolloClient = ApolloClient(networkTransport: requestChainTransport, store: store)
  }

  func taskList(limit: Int, cursor: String?, ignoreCache: Bool = false) async throws -> [TaskListItem] {
    try await withCheckedThrowingContinuation { con in
      apolloClient.fetch(
        query: WorkrecGraphql.TaskListQuery(limit: limit, cursor: nil),
        cachePolicy: ignoreCache ? .fetchIgnoringCacheData : .returnCacheDataElseFetch
      ) { result in
        self.resume(result: result, continuation: con) {
          $0.tasks.edges.map {
            TaskListItem(
              id: $0.node.id,
              title: $0.node.title,
              state: TaskState(rawValue: $0.node.state)!,
              totalWorkingTime: $0.node.totalWorkingTime,
              lastStartTime: ISO8601DateFormatter().date(from: $0.node.lastWork.startTime) ?? Date(timeIntervalSince1970: 0)
            )
          }
        }
      }
    }
  }

  func createTask(title: String) async throws {
    try await withCheckedThrowingContinuation { con in
      apolloClient.perform(mutation: WorkrecGraphql.CreateTaskMutation(title: title)) { result in
        self.resume(result: result, continuation: con) { _ in }
      }
    }
  }

  func startWorkOnTask(id: String, timestamp: Date) async throws {
    try await withCheckedThrowingContinuation { con in
      apolloClient.perform(mutation: WorkrecGraphql.StartWorkOnTaskMutation(taskId: id, timestamp: timestamp.ISO8601Format())) { result in
        self.resume(result: result, continuation: con) { _ in }
      }
    }
  }

  func stopWorkOnTask(id: String, timestamp: Date) async throws {
    try await withCheckedThrowingContinuation { con in
      apolloClient.perform(mutation: WorkrecGraphql.StopWorkOnTaskMutation(taskId: id, timestamp: timestamp.ISO8601Format())) { result in
        self.resume(result: result, continuation: con) { _ in }
      }
    }
  }

  private func resume<T, R>(result: Result<GraphQLResult<T>, Error>, continuation: CheckedContinuation<R, Error>, convert: (T) -> R) {
    switch result {
    case .success(let response):
      if let errors = response.errors {
        continuation.resume(throwing: ApiError(message: errors[0].errorDescription ?? ""))
        return
      }

      continuation.resume(returning: convert(response.data!))

    case .failure(let error):
      continuation.resume(throwing: error)
    }
  }
}

class UserManagementInterceptor: ApolloInterceptor {
  let provideAuthToken: () -> String?

  init(provideAuthToken: @escaping () -> String?) {
    self.provideAuthToken = provideAuthToken
  }

  enum UserError: Error {
    case noUserLoggedIn
  }

  /// Helper function to add the token then move on to the next step
  private func addTokenAndProceed<Operation: GraphQLOperation>(
    _ token: String,
    to request: HTTPRequest<Operation>,
    chain: RequestChain,
    response: HTTPResponse<Operation>?,
    completion: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void
  ) {
    request.addHeader(name: "Authorization", value: "Bearer \(token)")
    chain.proceedAsync(
      request: request,
      response: response,
      completion: completion)
  }

  func interceptAsync<Operation: GraphQLOperation>(
    chain: RequestChain,
    request: HTTPRequest<Operation>,
    response: HTTPResponse<Operation>?,
    completion: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void
  ) {
    guard let token = provideAuthToken() else {
      // In this instance, no user is logged in, so we want to call
      // the error handler, then return to prevent further work
      chain.handleErrorAsync(
        UserError.noUserLoggedIn,
        request: request,
        response: response,
        completion: completion
      )
      return
    }

    self.addTokenAndProceed(
      token,
      to: request,
      chain: chain,
      response: response,
      completion: completion
    )
  }
}

struct NetworkInterceptorProvider: InterceptorProvider {
  let provideAuthToken: () -> String?

  // These properties will remain the same throughout the life of the `InterceptorProvider`, even though they
  // will be handed to different interceptors.
  private let client: URLSessionClient

  private let store: ApolloStore

  init(client: URLSessionClient, store: ApolloStore, provideAuthToken: @escaping () -> String?) {
    self.client = client
    self.store = store
    self.provideAuthToken = provideAuthToken
  }

  func interceptors<Operation: GraphQLOperation>(for operation: Operation) -> [ApolloInterceptor] {
    return [
      MaxRetryInterceptor(),
      CacheReadInterceptor(store: self.store),
      UserManagementInterceptor(provideAuthToken: provideAuthToken),
      NetworkFetchInterceptor(client: self.client),
      ResponseCodeInterceptor(),
      JSONResponseParsingInterceptor(),
      AutomaticPersistedQueryInterceptor(),
      CacheWriteInterceptor(store: self.store),
    ]
  }
}

File: web-client/app/api-client.ts
export type TaskState = 'not_started' | 'in_progress' | 'paused' | 'completed'

export interface TaskListItem {
  id: string
  title: string
  totalWorkingTime: number
  state: TaskState
  lastStartTime: Date
}

export interface WorkSession {
  id: string
  startTime: Date
  endTime: Date
}

export interface TaskDetail {
  id: string
  title: string
  totalWorkingTime: number
  state: TaskState
  workSessions: WorkSession[]
}

export function taskDetailFromJson(obj: any): TaskDetail {
  return {
    id: obj.id,
    title: obj.title,
    totalWorkingTime: obj.totalWorkingTime,
    state: obj.state,
    workSessions: obj.workSessions,
  }
}

export function workSessionFromJson(obj: any): WorkSession {
  return {
    id: obj.id,
    startTime: obj.startTime ? new Date(obj.startTime) : new Date(0),
    endTime: obj.endTime ? new Date(obj.endTime) : new Date(0),
  }
}

export function taskListItemFromJson(obj: any): TaskListItem {
  return {
    id: obj.id,
    title: obj.title,
    totalWorkingTime: obj.totalWorkingTime,
    state: obj.state,
    lastStartTime: obj.lastStartTime
      ? new Date(obj.lastStartTime)
      : new Date(0),
  }
}

export async function fetchTaskList(
  sessionCookie: string,
  limit: number,
  cursor: string = ''
): Promise<TaskListItem[]> {
  const query = `
  query TaskList($limit: Int!, $cursor: String) {
    tasks(limit: $limit, cursor: $cursor) {
      edges {
        node {
          id
          state
          title
          totalWorkingTime
          lastWork {
            startTime
          }
        }
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }
  }
`

  const variables = { limit, cursor }

  return await fetchGraphql({ query, variables, sessionCookie }).then(
    (result) => {
      return result.data.tasks.edges.map((edge: any) => {
        return taskListItemFromJson({
          ...edge.node,
          lastStartTime: edge.node.lastWork?.startTime,
        })
      })
    }
  )
}

export async function fetchTaskDetail(
  sessionCookie: string,
  taskId: string,
  workSessionLimit: number,
  workSessionCursor?: string
): Promise<TaskDetail> {
  const query = `
  query TaskDetail($taskId: ID!, $workSessionLimit: Int!, $workSessionCursor: String) {
    task(id: $taskId) {
      id
      state
      title
      totalWorkingTime
      workSessions(limit: $workSessionLimit, cursor: $workSessionCursor) {
        edges {
          node {
            id
            startTime
            endTime
          }
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    }
  }
`

  const variables = { taskId, workSessionLimit, workSessionCursor }

  return await fetchGraphql({ query, variables, sessionCookie }).then(
    (result) => {
      return taskDetailFromJson({
        ...result.data.task,
        workSessions: result.data.task.workSessions.edges.map((edge: any) =>
          workSessionFromJson(edge.node)
        ),
      })
    }
  )
}

export async function createTask(
  sessionCookie: string,
  title: string
): Promise<void> {
  const query = `
  mutation CreateTask($title: String!) {
    createTask(title: $title) {
      id
      state
      title
      totalWorkingTime
      lastWork {
        id
        startTime
        endTime
        workingTime
      }
    }
  }
`
  const variables = { title }

  await fetchGraphql({ query, variables, sessionCookie })
}

export async function startWorkOnTask(
  sessionCookie: string,
  id: string
): Promise<void> {
  const query = `
  mutation StartWorkOnTask($taskId: String!, $timestamp: DateTime!) {
    startWorkOnTask(taskId: $taskId, timestamp: $timestamp) {
      id
      state
      title
      totalWorkingTime
      lastWork {
        id
        startTime
        endTime
        workingTime
      }
    }
  }
`

  const variables = { taskId: id, timestamp: new Date().toISOString() }
  await fetchGraphql({ query, variables, sessionCookie })
}

export async function stopWorkOnTask(
  sessionCookie: string,
  id: string
): Promise<void> {
  const query = `
  mutation StopWorkOnTask($taskId: String!, $timestamp: DateTime!) {
    stopWorkOnTask(taskId: $taskId, timestamp: $timestamp) {
      id
      state
      title
      totalWorkingTime
      lastWork {
        id
        startTime
        endTime
        workingTime
      }
    }
  }
`

  const variables = { taskId: id, timestamp: new Date().toISOString() }
  await fetchGraphql({ query, variables, sessionCookie })
}

export async function addWorkSession(
  sessionCookie: string,
  taskId: string,
  startTime: Date,
  endTime: Date
): Promise<void> {
  const query = `
  mutation AddWorkSession($taskId: String!, $startTime: DateTime!, $endTime: DateTime!) {
  addWorkSession(taskId: $taskId, startTime: $startTime, endTime: $endTime) {
      id
      startTime
      endTime
    }
  }
`

  const variables = { taskId, startTime, endTime }
  await fetchGraphql({ query, variables, sessionCookie })
}

export async function updateWorkSession(
  sessionCookie: string,
  id: string,
  startTime: Date,
  endTime: Date
): Promise<void> {
  const query = `
  mutation EditWorkSession($workSessionId: String!, $startTime: DateTime!, $endTime: DateTime!) {
  editWorkSession(workSessionId: $workSessionId, startTime: $startTime, endTime: $endTime) {
      id
      startTime
      endTime
    }
  }
`

  const variables = { workSessionId: id, startTime, endTime }
  await fetchGraphql({ query, variables, sessionCookie })
}

async function fetchGraphql({
  sessionCookie,
  query,
  variables,
}: {
  sessionCookie: string
  query: string
  variables: any
}): Promise<any> {
  const url = `${process.env.API_ORIGIN}/graphql`
  return await fetch(url, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${sessionCookie}`,
    },
    method: 'POST',
    body: JSON.stringify({
      query,
      variables,
    }),
  })
    .then((res) => res.json())
    .then((result) => {
      if (result.errors) {
        console.error(result)
        throw result.errors
      }
      return result
    })
}
